include/albatross/models/sparse_gp.h

```diff
@@ -0,0 +1,242 @@
+ * may mean diagonal or block diagonal.  Regardless we end up with O(m^2n)
+ * complexity instead of O(n^3) of direct Gaussian processes.  (Note that in [2]
+ * S is called sigma and A is lambda.)
+ *
+ *  Of course, the implementation details end up somewhat more complex in order
+ * to improve numerical stability.  A few great resources were heavily used to
+ * get those deails straight:
+ *
+ *     - https://bwengals.github.io/pymc3-fitcvfe-implementation-notes.html
+ *     - https://github.com/SheffieldML/GPy see fitc.py
+ */
+template <typename CovFunc, typename InducingPointStrategy,
+          typename IndexingFunction>
+class SparseGaussianProcessRegression
+    : public GaussianProcessBase<
+          CovFunc, SparseGaussianProcessRegression<
+                       CovFunc, InducingPointStrategy, IndexingFunction>> {
+
+public:
+  using Base = GaussianProcessBase<
+      CovFunc, SparseGaussianProcessRegression<CovFunc, InducingPointStrategy,
+                                               IndexingFunction>>;
+
+  SparseGaussianProcessRegression() : Base(){};
+  SparseGaussianProcessRegression(CovFunc &covariance_function)
+      : Base(covariance_function){};
+  SparseGaussianProcessRegression(
+      CovFunc &covariance_function,
+      InducingPointStrategy &inducing_point_strategy_,
+      IndexingFunction &independent_group_indexing_function_,
+      const std::string &model_name)
+      : Base(covariance_function, model_name),
+        inducing_point_strategy(inducing_point_strategy_),
+        independent_group_indexing_function(
+            independent_group_indexing_function_){};
+  SparseGaussianProcessRegression(CovFunc &covariance_function,
+                                  const std::string &model_name)
+      : Base(covariance_function, model_name){};
+
+  template <typename FeatureType>
+  auto _fit_impl(const std::vector<FeatureType> &features,
+                 const MarginalDistribution &targets) const {
+    static_assert(std::is_same<IndexingFunction, LeaveOneOut>::value,
+                  "Only FITC is currently implemented, but by allowing "
+                  "arbitrary groups this could turn into PITC");
+
+    // Determine the set of inducing points, u.
+    auto u = inducing_point_strategy(features);
+
+    Eigen::Index n = targets.mean.size();
+    Eigen::Index m = static_cast<Eigen::Index>(u.size());
+
+    // Create the covariance matrices we need.
+    Eigen::VectorXd K_ff_diag = this->covariance_function_.diagonal(features);
+    if (targets.has_covariance()) {
+      K_ff_diag += targets.covariance.diagonal();
+    }
+    Eigen::MatrixXd K_fu = this->covariance_function_(features, u);
+    Eigen::MatrixXd K_uu = this->covariance_function_(u);
+
+    auto K_uu_llt = K_uu.llt();
+    // P is such that:
+    //     Q_ff = K_fu K_uu^-1 K_uf
+    //          = K_fu L^-T L^-1 K_uf
+    //          = P P^T
+    Eigen::MatrixXd P = K_uu_llt.matrixL().solve(K_fu.transpose());
```

Comment:
Either the comment or the variable name should change, this is P^T.

Reply:
nice catch.